honeycomb/honeycomb.go

```diff
@@ -29,83 +31,287 @@ import (
+				return err
+			}
+			c.staticFields[name] = value
+		}
+		if c.dynamicFields != nil {
+			for name := range m {
+				delete(c.dynamicFields, name)
+			}
+		}
+		return nil
+	}
+}
+
+func validateDynamicField(name string, f func() interface{}) error {
+	if len(name) == 0 {
+		return errors.New("dynamic field name must not be empty")
+	}
+	if f == nil {
+		return fmt.Errorf("dynamic field %q must have a non-nil function", name)
+	}
+	return nil
+}
+
+// WithDynamicField adds a dynamic field with the given name to the
+// exporter. Any events published by this exporter will include a field with the
+// given name and a value supplied by invoking the corresponding function.
+//
+// This function replaces any field registered previously with the same name.
+func WithDynamicField(name string, f func() interface{}) ExporterOption {
+	return func(c *exporterConfig) error {
+		if err := validateDynamicField(name, f); err != nil {
+			return err
+		}
+		if c.dynamicFields == nil {
+			c.dynamicFields = make(map[string]func() interface{}, expectedDynamicFieldCount)
+		}
+		c.dynamicFields[name] = f
+		if c.staticFields != nil {
+			delete(c.staticFields, name)
+		}
+		return nil
+	}
+}
+
+// WithDynamicFields adds a set of dynamic fields to the exporter. Any events
+// published by this exporter will include fields pairing each name in the given
+// map with a value supplied by invoking the corresponding function.
+//
+// This function replaces any field registered previously with the same name.
+func WithDynamicFields(m map[string]func() interface{}) ExporterOption {
+	return func(c *exporterConfig) error {
+		count := len(m)
+		if count == 0 {
+			return nil
+		}
+		if c.dynamicFields == nil {
+			if count < expectedDynamicFieldCount {
+				count = expectedDynamicFieldCount
+			}
+			c.dynamicFields = make(map[string]func() interface{}, count)
+		}
+		for name, f := range m {
+			if err := validateDynamicField(name, f); err != nil {
+				return err
+			}
+			c.dynamicFields[name] = f
+		}
+		if c.staticFields != nil {
+			for name := range m {
+				delete(c.staticFields, name)
+			}
+		}
+		return nil
+	}
+}
+
+// WithAPIURL specifies the URL for the Honeycomb API server to which to send
+// events.
+//
+// If not specified, the default URL is https://api.honeycomb.io/.
+func WithAPIURL(url string) ExporterOption {
+	return func(c *exporterConfig) error {
+		// NB: libhoney.VerifyAPIKey parses this URL to make sure it's valid.
+		if len(url) == 0 {
+			return errors.New("API URL name must not be empty")
+		}
+		c.apiURL = url
+		return nil
+	}
+}
+
+// WithUserAgentAddendum specifies additional HTTP user agent-related detail to
+// include in HTTP requests issued to send events to the Honeycomb API
+// server. This value is appended to the user agent value from the libhoney
+// library.
+//
+// If not specified, the default value is "Honeycomb-OpenTelemetry-exporter." If
+// specified as an empty string, no user agent details are appended.
```

Comment:
@paulosman, note this proposed change in behavior: If a caller says that they want *no* addendum, we don't set the `libhoney.UserAgentAddition` field at all. Is that acceptable?

Reply:
@seh I'd like to avoid that. Let's treat empty string as invalid input and use the default.

Reply:
Commit 0fc3da4bbf740517ba60c122e3ac1148c1889df0 makes it so.